coalib/settings/SectionFilling.py

```diff
@@ -28,19 +28,39 @@ def fill_settings(sections, acquire_settings, log_printer):
     local_bears = {}
     global_bears = {}
 
+    bears_by_section = {}
+    bear_dirs_by_section = {}
+
     for section_name, section in sections.items():
-        bear_dirs = section.bear_dirs()
-        bears = list(section.get("bears", ""))
-        section_local_bears, section_global_bears = collect_bears(
-            bear_dirs,
-            bears,
-            [BEAR_KIND.LOCAL, BEAR_KIND.GLOBAL],
-            log_printer)
+        bear_dirs_by_section[section_name] = section.bear_dirs()
+        bears_by_section[section_name] = list(section.get("bears", ""))
+
+    bears_to_collect = set(chain.from_iterable(bears_by_section.values()))
+    bear_dirs_to_collect = set(
+        chain.from_iterable(bear_dirs_by_section.values()))
+
+    all_local_bears, all_global_bears = collect_bears(
+        bear_dirs_to_collect,
+        bears_to_collect,
+        [BEAR_KIND.LOCAL, BEAR_KIND.GLOBAL],
+        log_printer)
+    all_bears = list(chain(all_local_bears, all_global_bears))
+
+    for section_name, section in sections.items():
+        section_local_bears = []
+        section_global_bears = []
+
+        for bear in all_bears:
+            if bear.name in bears_by_section[section_name]:
```

Comment:
hm this could fail if the bear.name is not the same as the filename! What do we do about this? Any alternative?